### PR TITLE
Imports netsuite employees from gateway

### PR DESCRIPTION
**Pull request by sarony**
- TODO: meeds sad path of importing unsuccessfully

### CODE REVIEW

File: app/models/import_result.rb [hunk 1]
```diff
@@ -15,10 +15,9 @@ def to_s(format = "%{candidate}: %{result}\n")
     end.join
   end
 
-  def to_a 
-    results.map do |candidate, result| 
-      {candidate: candidate, 
-       result: result}
+  def to_a
+    results.map do |candidate, result|
+      { candidate: candidate, result: result }
```
Comment (hrs) [inline app/models/import_result.rb]: I might opt for:
```ruby
{
  candidate: candidate,
  result: result,
}
```

File: app/models/netsuite/attribute_mapper.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
+module Netsuite
+  class AttributeMapper
+    def call(netsuite_employee)
+      {
+        first_name: netsuite_employee["first_name"],
+        last_name: netsuite_employee["last_name"],
+        email: netsuite_employee["email"],
+        user_status: netsuite_status[netsuite_employee["is_inactive"]],
```
Comment (hrs) [inline app/models/netsuite/attribute_mapper.rb]: Would this read better as a short method?

File: app/models/netsuite/import.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+      )
+    end
+
+    private
+
+    attr_reader :namely_importer, :user
+    delegate :namely_connection, to: :user
+
+    def netsuite_employees
+      result ||= RestClient.get(ENV["NETSUITE_GATEWAY_URL"])
```
Comment (hrs) [inline app/models/netsuite/import.rb]: This isn't an instance variable, so nothing's being memoized here. If you want memoization, I'd probably do something like:
```ruby
def netsuite_employees
  @netsuite_employees ||= fetch_netsuite_employees
end

def fetch_netsuite_employees
  results = RestClient.get(ENV["NETSUITE_GATEWAY_URL"])
  JSON.parse(results)
end
```

File: app/views/netsuite_imports/create.html.erb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+<%= render 'subnav' %>
```
Comment (hrs) [inline app/views/netsuite_imports/create.html.erb]: Double-quotes, plz!

File: spec/fixtures/api_responses/netsuite_employee.json [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{
+  "first_name" : "Brandy",
+  "last_name" : "Dickens",
+  "email" : "bdickens@ramsey.com",
+  "internal_id" : "912",
+  "is_inactive" : "false",
+  "gender" : "female"
```
Comment (hrs) [inline spec/fixtures/api_responses/netsuite_employee.json]: Trailing comma!

File: app/views/netsuite_imports/create.html.erb (hunk 4 above)
Comment (cored) [inline app/views/netsuite_imports/create.html.erb]: Just a question, what would be the advantage of using double quotes in
here? It's easier for me ti type a single quote and she is not using
interpolation so why bother?

File: spec/fixtures/api_responses/netsuite_employee.json (hunk 5 above)
Comment (hrs) [inline spec/fixtures/api_responses/netsuite_employee.json]: Hmm, wait, this is JSON. I'd expect that we'd still favor trailing commas there, though.

File: spec/models/netsuite/attribute_mapper_spec.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
+require "rails_helper"
+
+describe Netsuite::AttributeMapper do
+  it "transforms a list of employees into an array of Hashes appropriate for
+  the Namely API" do
```
Comment (hrs) [inline spec/models/netsuite/attribute_mapper_spec.rb]: These should probably be on the same line.

File: spec/models/netsuite/import_spec.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+        status = netsuite_import.import
+
+        expect(status).to eq expected_status
+        expect(namely_importer).to have_received(:import).with(
+          recent_hires: [netsuite_employee],
+          namely_connection: user.namely_connection,
+          attribute_mapper: instance_of(Netsuite::AttributeMapper),
+        )
+
```
Comment (hrs) [inline spec/models/netsuite/import_spec.rb]: No newline here; both blocks are part of the "test" section.

File: spec/features/user_imports_netsuite_employees_spec.rb [hunk 8]
```diff
@@ -0,0 +1,35 @@
+require "rails_helper"
+
+feature "User imports netsuite employees" do
+  before do
+    stub_request(:get, "#{ api_host }/api/v1/profiles")
+      .with(query: {access_token: ENV['TEST_NAMELY_ACCESS_TOKEN'], limit: 'all'})
+      .to_return(status: 200, body: File.read("spec/fixtures/api_responses/empty_profiles.json"))
+  end
```
Comment (hrs) [inline spec/features/user_imports_netsuite_employees_spec.rb]: I might change this `before` block into a method (with a helpful name :grinning:), and then explicitly call it in the "setup" section of my test.

File: spec/features/user_imports_netsuite_employees_spec.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  let(:api_host) do
+    "%{protocol}://%{subdomain}.namely.com" % {
+      protocol: Rails.configuration.namely_api_protocol,
+      subdomain: ENV['TEST_NAMELY_SUBDOMAIN'],
+    }
+  end
```
Comment (hrs) [inline spec/features/user_imports_netsuite_employees_spec.rb]: I'd probably turn this `let` block into a function, too. It's only being used once, so the usual benefit of `let` (automatic memoization) doesn't help us here.

File: spec/features/user_imports_netsuite_employees_spec.rb [hunk 10]
```diff
@@ -0,0 +1,35 @@
+require "rails_helper"
+
+feature "User imports netsuite employees" do
+  before do
+    stub_request(:get, "#{ api_host }/api/v1/profiles")
```
Comment (hrs) [inline spec/features/user_imports_netsuite_employees_spec.rb]: I usually haven't seen us put whitespace inside of interpolations before—I'm used to `#{api_host}`, not `#{ api_host }`. When the guides talk about whitespace around braces, I think they're thinking of blocks, not interpolation (though I might be wrong...)

File: app/models/netsuite/attribute_mapper.rb [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    def namely_identifier_field
+      :netsuite_id
+    end
+
+    def identifier(netsuite_employee)
+      netsuite_employee["internal_id"]
+    end
+
+    def netsuite_status
+      { "false" => "active", "true" => "inactive" }
```
Comment (hrs) [inline app/models/netsuite/attribute_mapper.rb]: I think I'd like these pairs better on separate lines.

File: app/views/netsuite_imports/create.html.erb (hunk 4 above)
Comment (hrs) [inline app/views/netsuite_imports/create.html.erb]: A couple reasons:

* It's in accordance with our [style guide](https://github.com/thoughtbot/guides/tree/master/style/ruby), in which we "Prefer double quotes for strings." We adhered to that style guide throughout the project, so it's nice to keep the codebase consistent and clean.
* Our reasoning for that guideline was [discussed in an old PR](https://github.com/thoughtbot/guides/pull/174), but it boiled down to, "Using double-quotes everywhere improves consistency. It's one less bit of syntax we have to think about, and there's no appreciable performance hit."

Comment (cored) [inline app/views/netsuite_imports/create.html.erb]: Gotcha.

File: app/models/import_result.rb (hunk 1 above)
Comment (tabfugnic) [inline app/models/import_result.rb]: Would this just be for readability ?

Comment (hrs) [inline app/models/import_result.rb]: Yuppers.

File: spec/fixtures/api_responses/netsuite_employee.json (hunk 5 above)
Comment (sarony) [inline spec/fixtures/api_responses/netsuite_employee.json]: It freaked out when I tried to add a trailing comma. Gave me a `JSON::ParserError`. From what I saw, don't think this is allowed for JSON.

Comment (hrs) [inline spec/fixtures/api_responses/netsuite_employee.json]: Ah, bummer. Oh well.

File: app/models/netsuite/attribute_mapper.rb [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    end
+
+    def netsuite_employee_status(netsuite_employee)
+      netsuite_status[netsuite_employee["is_inactive"]]
+    end
+
+    def netsuite_status
+      {
+        "false" => "active",
+        "true" => "inactive"
```
Comment (hrs) [inline app/models/netsuite/attribute_mapper.rb]: Trailing comma, please!

File: app/models/import_result.rb [hunk 13]
```diff
@@ -15,10 +15,12 @@ def to_s(format = "%{candidate}: %{result}\n")
...
 
-  def to_a 
-    results.map do |candidate, result| 
-      {candidate: candidate, 
-       result: result}
+  def to_a
+    results.map do |candidate, result|
+      {
+        candidate: candidate,
+        result: result
```
Comment (hrs) [inline app/models/import_result.rb]: Trailing comma, please!

File: app/views/netsuite_imports/create.html.erb [hunk 14]
```diff
@@ -0,0 +1,6 @@
+<%= render "subnav" %>
+
+<h1>Netsuite</h1>
+<% if @netsuite_import %>
```
Comment (hrs) [inline app/views/netsuite_imports/create.html.erb]: I might say `if @netsuite_import.present?`. 

Also, should there be an `else` condition here?

File: spec/features/user_imports_netsuite_employees_spec.rb [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+  def stubbed_netsuite_gateway_request
+    stub_request(:get, ENV["NETSUITE_GATEWAY_URL"]).
+      to_return(body: [netsuite_employee].to_json)
+  end
+
+  def stubbed_namely_get_request
+    stub_request(:get, "#{api_host}/api/v1/profiles").
+      with(
+        query: {
+          access_token: ENV['TEST_NAMELY_ACCESS_TOKEN'],
```
Comment (hrs) [inline spec/features/user_imports_netsuite_employees_spec.rb]: Double quotes!

File: spec/features/user_imports_netsuite_employees_spec.rb [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+          limit: 'all',
```
Comment (hrs) [inline spec/features/user_imports_netsuite_employees_spec.rb]: Double quotes!

File: Gemfile [hunk 17]
```diff
@@ -8,22 +8,23 @@ gem "coffee-rails"
 gem "delayed_job_active_record"
 gem "email_validator"
 gem "flutie"
+gem "font-awesome-rails"
 gem "high_voltage"
 gem "i18n-tasks"
 gem "jquery-rails"
 gem "namely", github: "namely/ruby-client"
 gem "neat", "~> 1.5.1"
+gem "netsuite", github: "tabfugnic/netsuite"
```
Comment (tabfugnic) [inline Gemfile]: Do you need to have this here?

File: app/controllers/jobvite_imports_controller.rb [hunk 18]
```diff
@@ -1,7 +1,7 @@
 class JobviteImportsController < ApplicationController
   def create
     jobvite_import = Jobvite::Import.new(current_user)
-    @jobvite_imports_presenter = Jobvite::ImportsPresenter.
+    @jobvite_imports_presenter = NamelyImportsPresenter.
```
Comment (tabfugnic) [inline app/controllers/jobvite_imports_controller.rb]: Let's change the name of this more to the `ImportsPresenter`

File: app/views/jobvite_imports/create.html.erb [hunk 19]
```diff
@@ -48,7 +48,7 @@
       </tr>
     </thead>
     <tbody>
-      <% @jobvite_imports_presenter.not_imported_candidates.each do |import_result| %>
+      <% @jobvite_imports_presenter.not_imported.each do |import_result| %>
```
Comment (tabfugnic) [inline app/views/jobvite_imports/create.html.erb]: `not_imported` doesn't say anything to me about people or that it even resturs an array. I'd be happier to call these employees.

File: config/locales/en.yml [hunk 20]
```diff
@@ -44,15 +44,24 @@ en:
...
+      imported_successfully: "You have successfully imported the following:"
+      key_title: "Key"
+      not_imported_successfully: "You haven't import the following:"
       reason: "Reason"
 
+  netsuite_imports:
+    create:
+      email_address_title: "Email Address"
+      empty_successful_imports: "Currently there aren't any profiles to import"
+      employee_title: "Employee Name"
```
Comment (tabfugnic) [inline config/locales/en.yml]: :abc:

File: config/locales/en.yml [hunk 21]
```diff
@@ -44,15 +44,24 @@ en:
...
     create:
-      imported_successfully: "You have successfully imported the following:"
-      empty_successful_imports: "Currently there aren't any profiles to import"
-      not_imported_successfully: "You haven't import the following:"
-      employee_title: "Employee Name"
-      key_title: "Key"
       contact_number_title: "Contact Number"
       email_address_title: "Email Address"
+      employee_title: "Employee Name"
+      empty_successful_imports: "Currently there aren't any profiles to import"
```
Comment (tabfugnic) [inline config/locales/en.yml]: :abc:

File: app/views/jobvite_imports/create.html.erb (hunk 19 above)
Comment (sarony) [inline app/views/jobvite_imports/create.html.erb]: Even though in the jobvite context, they're not necessarily employees yet?

File: Gemfile (hunk 17 above)
Comment (sarony) [inline Gemfile]: True

File: spec/models/netsuite/attribute_mapper_spec.rb [hunk 22]
```diff
@@ -0,0 +1,29 @@
+require "rails_helper"
+
+describe Netsuite::AttributeMapper do
+  describe ".call" do
+    it "transforms a Netsuite employee into hash for the Namely API" do
+      netsuite_employee = double(
```
Comment (tabfugnic) [inline spec/models/netsuite/attribute_mapper_spec.rb]: I feel like the hope here is that this comes down as a candidate no? The api would not send information as the netsuite::employee

Comment (sarony) [inline spec/models/netsuite/attribute_mapper_spec.rb]: When it gets to the `AttributeMapper`, it's already taken the employee information from the api and made a `Netsuite::Employee` for each employee. This takes place in `Netsuite::Import` line 23. Does that answer the question?

**Comment by tabfugnic** [comment]
Since this integration doesn't appear to be going through any time soon, I will close this for the time being and we can readdress that later.

**Pull request closed without merging**